add stack trace to device skip

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118113
* __->__ #118112

Log stack trace of offending cpu use if it causes a disabling of cudagraphs. Also refactoring disable_cudagraphs: bool, and disable_cudagraphs_reason: str -> Optional[str].



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

eellison committed:
add stack trace to device skip

[ghstack-poisoned]

bdhirsh commented on torch/_inductor/cudagraph_utils.py:
```diff
@@ -54,3 +54,35 @@ def check_for_mutation(
+    return None
+
+
+def check_multiple_devices(
```
probably a nit but: maybe rename to `check_multiple_devices_or_any_cpu_nodes`? IIRC, you can have a graph that's 100% cpu (single-device) but we want this fn to trip and not use cudagraphs

eellison replied:
Sure will do

bdhirsh reviewed (approved):
lgtm, nits

eellison committed:
Update on "add stack trace to device skip"


Log stack trace of offending cpu use if it causes a disabling of cudagraphs. Also refactoring disable_cudagraphs: bool, and disable_cudagraphs_reason: str -> Optional[str].



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot merge
